cmd/snap-confine/snap-confine.c

```diff
@@ -347,11 +333,104 @@ int main(int argc, char **argv) {
+        // that supports AppArmor when the effective uid is root and the real
+        // id is non-root.  This protects against, for example, unprivileged
+        // users trying to leverage the snap-confine in the core snap to
+        // escalate privileges.
+        errno = 0;  // errno is insignificant here
+        die("snap-confine has elevated permissions and is not confined"
+            " but should be. Refusing to continue to avoid"
+            " permission escalation attacks\n"
+            "Please make sure that the snapd.apparmor service is enabled and started.");
+    }
+
+    sc_debug_capabilities("initial caps");
+
+    static const cap_value_t snap_confine_caps[] = {
+        CAP_DAC_OVERRIDE,     // poking around as a regular user
+        CAP_DAC_READ_SEARCH,  // same as above
+        CAP_SYS_ADMIN,        // mounts, unshare
+        CAP_SYS_CHROOT,       // pivot_root into a new root
+        CAP_CHOWN,            // file ownership
+        CAP_FOWNER,           // to create tmp dir with sticky bit
+        CAP_SYS_PTRACE,       // to inspect the mount namespace of PID1
+        CAP_SETUID,           // assume user identity
+        CAP_SETGID,           // and group identity, e.g. switching to root when running s-u-n
+    };
+
+    /* We may be invoking tools such as snap-update-ns or snap-discard which are
+     * executed in a forked process, the child can inherit at most these
+     * capabilities */
+    static const cap_value_t helper_tools_inheritable_caps[] = {
+        CAP_DAC_OVERRIDE,  // poking around as a regular user
+        CAP_SYS_ADMIN,     // mounts
+        CAP_CHOWN,         // file ownership
+    };
+
+    /* Capability setup:
+     * 1. Permitted caps are obtained from file.
+     * 2. Restore those capabilities that we really need into the
+     *    "effective" set.
+     * 3. Capabilities needed by either us or by any of our child processes
+     *    need to be set into the "permitted" set.
+     * 4. Capabilities needed by our helper child processes need to be set
+     *    into the "permitted", "inheritable" and "ambient" sets.
+     *
+     * Before executing the snap application we'll drop all capabilities.
+     */
+    cap_t caps_privileged SC_CLEANUP(cap_free) = cap_get_proc();
+    if (caps_privileged == NULL) {
+        die("cannot obtain current caps");
+    }
+
+    if (cap_set_flag(caps_privileged, CAP_EFFECTIVE, SC_ARRAY_SIZE(snap_confine_caps), snap_confine_caps, CAP_SET) !=
+        0) {
+        die("cannot fill effective capability set");
+    }
+    /* inheritable caps when forking off to a helper tool */
+    if (cap_set_flag(caps_privileged, CAP_INHERITABLE, SC_ARRAY_SIZE(helper_tools_inheritable_caps),
+                     helper_tools_inheritable_caps, CAP_SET) != 0) {
+        die("cannot fill inheritable capability set");
+    }
+
+    cap_t caps_no_effective SC_CLEANUP(cap_free) = cap_dup(caps_privileged);
```

Comment:
this probably need a comment to define what they are/for

Reply:
added